### PR TITLE
[FIRRTL] limit useless locs in dedup

### PR DESCRIPTION
**Pull request by darthscsi**
Limit locations from .fir files when deduplicating modules.  These locations add little value.  This has the unfortunate side-effect of embedding special handling of location data based on filename string, but I don't see a better way right now.

End-to-end performance increase on a large core: 15%
Memory reduction on a large core: 50%

### CODE REVIEW

File: lib/Dialect/FIRRTL/Transforms/Dedup.cpp [hunk 1]
```diff
@@ -577,6 +577,51 @@ struct Equivalence {
 // Deduplication
 //===----------------------------------------------------------------------===//
 
+static Location mergeLoc(MLIRContext *context, Location to, Location from) {
+  //    llvm::errs() << "TESTING \n";
+  //    to->dump();
+  //    from->dump();
+  //    llvm::errs() << "\n";
```
Comment (youngar) [inline lib/Dialect/FIRRTL/Transforms/Dedup.cpp:584]: extra debug code

File: lib/Dialect/FIRRTL/Transforms/Dedup.cpp [hunk 2]
```diff
@@ -577,6 +577,51 @@ struct Equivalence {
...
+    // metadata or the metadata is the same as the top level metadata.
+    if (auto fusedLoc = loc.dyn_cast<FusedLoc>()) {
+      // UnknownLoc's have already been removed from FusedLocs so we can
+      // simply add all of the internal locations.
+      for (auto loc : fusedLoc.getLocations()) {
+        if (FileLineColLoc fileLoc = loc.dyn_cast<FileLineColLoc>()) {
+          if (fileLoc.getFilename().strref().endswith(".fir")) {
+            ++seenFIR;
+            if (seenFIR > 8)
+              continue;
```
Comment (youngar) [inline lib/Dialect/FIRRTL/Transforms/Dedup.cpp:601]: could return `FusedLoc::get(context, seen);`

File: lib/Dialect/FIRRTL/Transforms/Dedup.cpp [hunk 3]
```diff
@@ -577,6 +577,51 @@ struct Equivalence {
...
+          }
+        }
+        decomposedLocs.insert(loc);
+      }
+      continue;
+    }
+    // Otherwise, only add known locations to the set.
+    if (!loc.isa<UnknownLoc>())
+      decomposedLocs.insert(loc);
```
Comment (youngar) [inline lib/Dialect/FIRRTL/Transforms/Dedup.cpp:610]: Probably should add the "fir check" here as well

**Review by youngar — APPROVED** [review]
Might want to add some tests

File: lib/Dialect/FIRRTL/Transforms/Dedup.cpp (hunk 1 above)
Comment (darthscsi) [reply lib/Dialect/FIRRTL/Transforms/Dedup.cpp]: forgot to save before the last push.

File: lib/Dialect/FIRRTL/Transforms/Dedup.cpp (hunk 3 above)
Comment (darthscsi) [reply lib/Dialect/FIRRTL/Transforms/Dedup.cpp:619]: yea, doesn't matter too much, but might as well be consistent.

**Comment by seldridge** [comment]
I'm confirming that this fixes the issue we were seeing with a memory blowup in `Dedup` for an internal design. With this patch, memory doesn't budge during `Dedup` and this is some 5x faster on that design.

**Pull request merged**